### PR TITLE
Emit a relative path in -gseparate-dwarf

### LINKED ISSUE #12526: -gseparate-dwarf should embed a path relative to the Wasm binary
**Steps to reproduce:**

 - Create an arbitrary sample file `sample.c` and a directory `dist`.
 - Try to compile arbitrary `sample.c` with the following command:
   ```sh
   $ emcc sample.c -o dist/sample.html -g -gseparate-dwarf=dist/sample.dbg.wasm
   ```

**Expected result:**

1. The stripped Wasm file is generated in `dist/sample.wasm`
2. The debug version of the Wasm file is generated in `dist/sample.dbg.wasm`
3. The `external_debug_info` in `dist/sample.wasm` contains a relative path `sample.dbg.wasm` that points to the debug binary.

**Actual result:**

(1) and (2) hold, but the `external_debug_info` contains a path `dist/sample.dbg.wasm` - relative to the compilation working directory and not to the Wasm.

As such, when the debugger tries to resolve the path to the debug info, it constructs a path `dist/dist/sample.dbg.wasm` and fails to load it.

### PR DESCRIPTION
**Pull request by kripken**
Emit only the basename in the wasm to point to the debug wasm. This assumes
the files are alongside each other. (For more complex things, we already have
`SEPARATE_DWARF_URL`).

cc @RReverser @pfaffe @dschuff 

fixes #12526

### CODE REVIEW

**Comment by RReverser** [comment]
> This assumes
> the files are alongside each other. (For more complex things, we already have
> `SEPARATE_DWARF_URL`).

I don't know what `SEPARATE_DWARF_URL` is, but I don't think that assumption is correct, since, for better or worse, `-gseparate-dwarf=...` accepts an arbitrary filename.

Instead, we should probably produce a relative URL using the same logic as in source maps.

**Comment by kripken** [comment]
`SEPARATE_DWARF_URL` basically lets you specify exactly what URL to embed.

I'm not sure what you're suggesting. Relative to what, if not to the wasm file itself? It seems like it has to be the wasm file since that's where we actually emit it?

**Comment by RReverser** [comment]
> Relative to what, if not to the wasm file itself?

Yes, I _am_ suggesting relative to Wasm file itself. But it's not what was happening before the change nor what is happening in the PR, IIUC?

**Comment by kripken** [comment]
I'm confused...

Building to say `-o subdir/output.wasm` will emit `subdir/output.wasm` and `subdir/output.debug.wasm`. The two files are alongside each other. And the path to the debug one will be just the basename, `output.debug.wasm`. That is the proper path when one runs the wasm in the location it is in.

Do you mean something else by "relative path"? If so, what should the path to the debug one be in the example I just gave?

**Comment by RReverser** [comment]
> The two files are alongside each other.

I think I see where confusion is coming from. You're talking about argumentless version of `-gseparate-dwarf`, which, indeed, produces `subdir/output.debug.wasm` by default, but I'm talking about a more generic case since the parameter allows to accept an explicit filename, e.g.:

```shell
$ emcc input.c -o subdir/output.wasm -g -gseparate-dwarf=output.wasm.dbg
```

In this case, the debug file will not be produced alongside the `output.wasm` but in the same folder as `input.c`, and so the relative path should end up being encoded as `../output.wasm.dbg` but current PR won't do that.

Does this make sense?

**Comment by RReverser** [comment]
But also, until this thread I didn't even know about

> `SEPARATE_DWARF_URL` basically lets you specify exactly what URL to embed.

I assumed that, when filename is accepted, at least relative URL will be taken care of anyway, and I think many users will expect the same.

**Comment by kripken** [comment]
Thanks @RReverser ! Yes, we were thinking of different usages of the flag. There are probably too many, but for now, this PR should fix them all.

**Review by RReverser — APPROVED** [review]
LGTM, thanks!

File: tools/building.py
Comment (RReverser) [inline tools/building.py:1366]: Btw, in theory you don't need this separate code path - the one below with `relpath` should cover both cases (since basename is just a partial case of relpath). So the code could be simplified to something like:

```python
if wasm_file_with_dwarf is True:
  wasm_file_with_dwarf = wasm_file + '.debug.wasm'
if not embedded_path:
  embedded_path = os.path.relpath(wasm_file_with_dwarf, os.path.dirname(wasm_file))
```

**Review by dschuff — APPROVED** [review]
I think Ingvar is right, LGTM if we switch to something like that.

File: tools/building.py
Comment (kripken) [reply tools/building.py:1366]: Good idea, done.

**Comment by dschuff** [comment]
It looks like this is failing on windows on chromium CI: https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket.appspot.com/8865709443195525296/+/steps/Emscripten_testsuite__upstream__other_/0/stdout

**Comment by RReverser** [comment]
Lol that was literally my first thought after merging (that we forgot about `\` -> `/` replacement like we did in other places), but then I thought "nah, if it passed CI, it should be fine".

**Pull request merged**
